spec/models/course_spec.rb

```diff
@@ -142,9 +142,457 @@
+          @assignment_before_call.assign_attributes({ assignment_properties_attributes: { scanned_exam: 'false',
+                                                                                          repository_folder: 'TEST' } })
+        end
+        context 'when the assignment already exists and an attribute is changed' do
+          before(:each) do
+            @assignment_before_call.message = 'a'
+            @assignment_before_call.save
+            @assignment_before_call.reload
+
+            # change the state of the assignment locally to be compared with the updated one in the database
+            @assignment_before_call.message = 'b'
+
+            # desired list of assignment attributes
+            @assignment_values = Assignment::DEFAULT_FIELDS.map do |f|
+              @assignment_before_call.public_send(f)
+            end
+
+            # csv representation of the single assignment
+            @csv = @assignment_values.to_csv
+
+            test_course.upload_assignment_list('csv', @csv)
+          end
+
+          it 'should update only the attributes that were changed' do
+            assignment_before_call_attr = @assignment_before_call.attributes
+            assignment_after_call_attr = test_course.assignments.find(@assignment_before_call.id).attributes
+
+            # Remove the time updated in preparation for comparison
+            assignment_before_call_attr.delete_if { |k, _| k == 'updated_at' }
+            assignment_after_call_attr.delete_if { |k, _| k == 'updated_at' }
+
+            # adjust the precision of assignment_before_call to account for precision loss
+            # in csv to object conversion
+            assignment_before_call_attr['due_date'] = assignment_before_call_attr['due_date'].floor
+
+            # check if the assignment stored locally, vs the one edited using upload_assignment_list match
+            expect(assignment_after_call_attr).to eq(assignment_before_call_attr)
+          end
+          it 'should not create a new assignment' do
+            # check if the number of assignments stored has changed
+            expect(test_course.assignments.length).to eq(1)
+          end
+        end
+        context 'when the assignment is new to the database' do
+          it 'should set assignment_properties.repository_folder, token_period and ' \
+             'unlimited_tokens to pre-determined values' do
+            # setting properties to different values than the ones they are supposed to be changed to
+            @assignment_before_call.token_period = nil
+            @assignment_before_call.unlimited_tokens = nil
+
+            # desired list of assignment attributes
+            desired_attributes = Assignment::DEFAULT_FIELDS.map do |f|
+              @assignment_before_call.public_send(f)
+            end
+
+            # this is done to assure assignment_properties.repository_folder = row[0]
```

Comment:
This part shouldn't be necessary if you use the assignment factory (which will set the repository_folder for you).